lib/test_state.go

```diff
@@ -31,6 +33,118 @@ type TestRunState struct {
 	Runner  Runner // TODO: rename to something better, see type comment
 	RunTags *metrics.TagSet
 
+	GroupSummary *GroupSummary // TODO(@mstoykov): move and rename
+
 	// TODO: add other properties that are computed or derived after init, e.g.
 	// thresholds?
 }
+
+// GroupSummaryDescription is the description of the GroupSummary used to identify and ignore it
+// for the purposes of the cli descriptions.
+const GroupSummaryDescription = "Internal Group Summary output"
+
+// GroupSummary is an internal output implementation that facilitates the aggregation of
+// group and check metrics for the purposes of the end of test summary and the REST API
+type GroupSummary struct {
+	group   *Group // TODO(@mstoykov): move the whole type outside of lib later
+	ch      chan []metrics.SampleContainer
+	stopped chan struct{}
+	logger  logrus.FieldLogger
+}
+
+// NewGroupSummary returns new GroupSummary ready to be started.
```

Comment:
I kind of wanted to move the whole Group, Check and GroupSummary somewhere else, but then I decided that is quite a big change and will make the review even harder.

It also turned out to not be easy.

So if we have any good places to move them I would prefer to do them in a separate PR.

Reply:
Ideally, I would use this opportunity to move them to an `internal` package. This is my unique request, reflecting what the comment is mentioning.

Reply:
I am not against this - but that will create a much bigger change that will make following what happens ever harder. So I will prefer to keep it in a separate PR.

Not certain `internal` is needed as well so :shrug: on that particular part. But it is probably a good usage as well.

Reply:
Yeah, especially with the v1.0 on the road, would be great to start considering hiding internal details behind an `internal` package, to reduce the public surface ( 👋🏻 say hello to [Hyrum's Law](https://www.hyrumslaw.com/)).

That said, I fully agree on doing so on a separate PR, to avoid making this even harder to read, follow and review.